src/storage/src/source/mod.rs

```diff
@@ -806,169 +866,640 @@ where
+            // Emit batches more frequently than we mint new timestamps. We're
+            // hoping that most of the batches that we emit will end up making
+            // it into the freshly minted bindings when remap_operator ticks.
+            let tick_interval = timestamp_frequency / 5;
 
-            let mut ts_upper = Antichain::from_elem(Timestamp::minimum());
             let mut untimestamped_messages = HashMap::<_, Vec<_>>::new();
             let mut non_definite_errors = vec![];
+
+            let mut last_tick_time = Instant::now();
+
+            // Keep track of how many messages we emitted before checking if we
+            // should tick. This avoids doing the costly checks for every
+            // message.
+            let mut messages_in_batch = 0;
+
             loop {
-                // TODO(guswyn): move lots of this out of the macro so rustfmt works better
-                tokio::select! {
-                    // N.B. This branch is cancel-safe because `next` only borrows the underlying stream.
-                    item = source_stream.next() => {
-                        match item {
-                            Some(Ok(message)) => {
-                                // Note that this
-                                // 1. Requires that sources that produce `InProgress` messages
-                                //    ALWAYS produce a `Finalized` for the final message.
-                                // 2. Requires that sources that produce `InProgress` messages
-                                //    NEVER produces messages at offsets below the most recent
-                                //    `Finalized` message.
-                                let is_final = matches!(message, SourceMessageType::Finalized(_));
-                                match message {
-                                    SourceMessageType::Finalized(message) | SourceMessageType::InProgress(message) => {
-                                        let pid = message.partition.clone();
-                                        let offset = message.offset;
-                                        // advance the _offset_ frontier if this the final message for that offset
-                                        if is_final {
-                                            *source_upper.entry(pid.clone()).or_default() = offset + 1;
-                                        }
-                                        untimestamped_messages.entry(pid).or_default().push((message, offset));
-                                    }
-                                    SourceMessageType::SourceStatus(update) => {
-                                        if let Some(healthchecker) = &mut healthchecker {
-                                            healthchecker.update_status(update).await;
-                                        }
-                                    }
+                let item = source_reader.get_next_message();
```

Comment:
Hm, we can't call `get_next_message` here because its default implementation always returns `Pending` (I should have probably made it panic instead) and doesn't delegate to `SourceReader::next` meaning that any source that only implements `next` will not work. We don't have any source that does this at the moment but it's a step back if we still want to deprecate `get_next_message`.